File: 2023/6/0615/linkedList.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next


class linkedList:
    def __init__(self, data):
        self.head = Node(data)

    def add(self, data):
        if self.head == '':
            self.head = Node(data)
        else:
            node = self.head
            while node.next:
                node = node.next
            node.next = Node(data)

    def desc(self):
        node = self.head
        while node:
            print(node.data)
            node = node.next

    def delete(self, data):
        if self.head == '':
            return

        if self.head.data == data:
            tmp = self.head
            self.head = self.head.next
            del tmp

        else:
            node = self.head
            while node.next:
                if node.next.data == data:
                    tmp = node.next
                    node.next = node.next.next
                    del tmp
                else:
                    node = node.next

    def search(self, data):
        node = self.head
        while node:
            if node.data == data:
                return node
            else:
                node = node.next

File: 2024/8/12/cpp/LinkedList.cpp
#include <iostream>
struct Node
{
    int data;
    Node *next;
    Node(int data) : data(data), next(nullptr) {};
};

class LinkedList
{
private:
    Node *head;

public:
    LinkedList() : head(nullptr) {};

    void append(int data)
    {
        Node *newNode = new Node(data);
        if (head == nullptr)
        {
            head = newNode;
            return;
        }
        Node *temp = head;
        while (temp->next){
            temp = temp->next;
        }
        temp->next = newNode;
        return;
    };

    void print()
    {
        if (head == nullptr) return;
        Node *temp = head;
        while (temp)
        {
            std::cout << temp->data << " ";
            temp=temp->next;
        }
        std::cout << std::endl;
    };

    void deleteNode(int key)
    {
        if (head == nullptr) return;
        Node *temp = head;
        Node *prev = nullptr;
        
        while(temp != nullptr && temp->data != key){
            prev = temp;
            temp=temp->next;
        }
        if (temp == nullptr) return;
        prev->next = temp->next;
        delete temp;
    }

    ~LinkedList()
    {
        Node *temp = head;
        while (temp)
        {
            Node *next = temp->next;
            delete temp;
            temp = next;
        }
    };
};

int main()
{
    LinkedList ll = LinkedList();
    ll.append(1);
    ll.print();
    ll.append(2);
    ll.print();
    ll.append(3);
    ll.print();
    ll.append(4);
    ll.print();
    ll.append(5);
    ll.print();
    ll.deleteNode(3);
    ll.print();
    ll.deleteNode(5);
    ll.print();
    return 0;
}